Add jvm_index Option to Coursier Subsystem

NGustafson opened the pull request:
Fixes #20260 

Adds the jvm_index option, with the default value set as the current default for Coursier.  The help message for the option is paraphrased from the description of the Coursier argument here https://get-coursier.io/docs/cli-java#jvm-index.

tdyas commented on src/python/pants/jvm/jdk_rules.py:
```diff
@@ -216,6 +216,8 @@ async def prepare_jdk_environment(
     else:
         coursier_jdk_option = shlex.quote(f"--jvm={version}")
 
+    jvm_index = f"--jvm-index {coursier.jvm_index}"
```
Does this work with the space instead of a `=`? The arguments are passed via an array of strings, thus I would think this should either be a single argument form like `f"--jvm-index={coursier.jvm_index}"` or should be two arguments `["--jvm-index", coursier.jvm_index]` and then use *jvm_index to insert it into the list later in the function.

tdyas replied:
Also, would it be better to not pass this argument in the default case?

NGustafson replied:
It seems to be working as expected with the space.  The example usage for the argument in the Coursier documentation looks like this:
```
cs java --available --jvm-index https://url/of/your/index.json
```
So I chose to use a space.  I can check if it works with `=` as well if it would be better to have consistency with how the version argument is formatted in line 217.

I like the idea of not passing the argument by default.  In that case would it be better for the default value of the option to be an empty string?

huonw replied:
Ah, I think the space does work as is, but this is formatted into a shell script directly (and thus the space works because it is relying on shell expansion/argument parsing).

Thus, the space working is revealing a deeper problem: it should be created with `shlex.quote` so that it behaves properly if the URL contains any characters that are special to a shell (if not, special characters like `;` will result in at best incorrect results, and at worse shell-injection vulnerabilities).

Thus, I think it should either, look similar to the `coursier_jdk_option` assignment on the previous line:

```suggestion
    jvm_index = shlex.quote(f"--jvm-index={coursier.jvm_index}")
```

(NB. the `=` then becomes required, or else it's like calling `cs java --available '--jvm-index https://example.com'`, which doesn't seem to work for me.)

Or, with the two argument version:

```suggestion
    jvm_index = ["--jvm-index", shlex.quote(coursier.jvm_index)]
```

kaos replied:
I think the shell quoting ought to be applied once, later. So this issue would've been caught rather than sprinkling with quoting calls.

tdyas replied:
> I think the shell quoting ought to be applied once, later (to all arguments). So this issue would've been caught rather than sprinkling with quoting calls.

I agree. Quoting with `shlex.quote` when we construct the final command will ensure less chance of problems in the future.

NGustafson replied:
Thank you for the suggestion.  I've looked into it a bit, but I'm not sure how to proceed with applying `shlex.quote` only once.  Quoting the entire command seems to cause errors when it's executed.

I was thinking I could apply `shlex.quote` to each value in `coursier_options` in a separate statement, or I could modify the prefixed function like this:
``` python
def prefixed(arg: str) -> str:
    if arg.startswith("__"):
        return f"${{PANTS_INTERNAL_ABSOLUTE_PREFIX}}{arg}"
    else:
        return shlex.quote(arg)
```
So at least any future additional options would be quoted when `optionally_prefixed_coursier_args` is defined.  Would that be sufficient?

NGustafson replied:
@kaos @tdyas Could either of you check my previous response?  I think I may have misunderstood your feedback, so I would like to confirm before making a change.

kaos replied:
How about quoting the args here:
https://github.com/pantsbuild/pants/blob/fe449a537423d6fa25ef6887e28094b1ec6959c5/src/python/pants/jvm/resolve/coursier_setup.py#L183-L189

```diff
    def args(self, args: Iterable[str], *, wrapper: Iterable[str] = ()) -> tuple[str, ...]:
        return (
            self.post_process_stderr,
            *wrapper,
            os.path.join(self.bin_dir, self.coursier.exe),
--          *args,
++          *map(shlex.quote, args),
        )
```

NGustafson replied:
Nice, I'll give that a try.  Thank you!

NGustafson commented:
Thank you for the feedback.
I've updated the argument to use `=` and `shlex.quote` as suggested.  I also tried to ensure the argument is only included when the option is defined, and changed the default value to an empty string.  Please let me know if there are any concerns with this approach.

huonw commented:
Ah, I only just noticed now that there's no test for this. Do others have a suggestion for how this might be tested?

NGustafson commented:
> Ah, I only just noticed now that there's no test for this. Do others have a suggestion for how this might be tested?

I've added a test where the `coursier-jvm-index` option points to an empty `index.json` file, which causes the specified jdk version to not be found.  Does that seem reasonable?

I'm wondering if I should add an extra check to ensure the specified jdk version is available when `coursier-jvm-index` isn't set, just in case that version is ever removed from the default coursier index.

NGustafson commented:
> I'm wondering if I should add an extra check to ensure the specified jdk version is available when `coursier-jvm-index` isn't set, just in case that version is ever removed from the default coursier index.

I decided to add this test case.
@huonw Could you check if the test I've added seems reasonable?

huonw commented on src/python/pants/jvm/resolve/coursier_setup.py:
```diff
@@ -185,7 +199,7 @@ def args(self, args: Iterable[str], *, wrapper: Iterable[str] = ()) -> tuple[str
             *wrapper,
             os.path.join(self.bin_dir, self.coursier.exe),
-            *args,
+            *map(shlex.quote, args),
```
Hm, thanks for trying this. I think technically this may not be quite right. This appears to be building the appropriate arguments for a single command, and (as it was), would've been safe/fine to pass to a command like `subprocess.run(coursier.args(...))` (or Pants' equivalent). By quoting here, that command would see the wrong argument, e.g. instead of something like `subprocess.run(["coursier", "has space"])` it would be `subprocess.run(["coursier", "'has space'"])` and the command would receive those quotes literally, unintentionally. This potentially happens at the following (although maybe the arguments rarely need quoting, so tests don't fail):

https://github.com/pantsbuild/pants/blob/100c4c7ba8e14c6d99663749b4c26eb2bf602fad/src/python/pants/jvm/resolve/coursier_setup.py#L239-L243


The special thing about what happens in `prepare_jdk_environment` is the invocation is splatted into a shell script as one long text string, rather than invoking it with a Python-ful API directly (and those APIs can preserve the separation between arguments in the list).

Thus, I think the best way to fix this is close to what you have but just  in the other file, i.e.

```diff
- optionally_prefixed_coursier_args = [prefixed(arg) for arg in coursier.args(coursier_options)]
+ optionally_prefixed_coursier_args = [shlex.quote(prefixed(arg)) for arg in coursier.args(coursier_options)]
```

and restore this back to how it was.

kaos replied:
🤦🏽 thanks for catching Huon..

NGustafson replied:
Applying `shlex.quote` after `prefixed` seems to be causing the environment variable appended by `prefixed` to be interpreted literally.  I get an error like.
```
/bin/bash: ${PANTS_INTERNAL_ABSOLUTE_PREFIX}__coursier/coursier_post_process_stderr.py: No such file or directory
```


Would changing the order like this cause any other issues?
```python
optionally_prefixed_coursier_args = [prefixed(shlex.quote(arg)) for arg in coursier.args(coursier_options)]
```

huonw replied:
Ah, yes, the quoting will indeed transform it to something like `'${PANTS_...}` intentionally, so that the `$` comes through literally, but that's not what we want. 

I think your proposal is close, but will cause issues if a `__` prefixed arg needs quoting, e.g. `__something here` would become `'__something here'`. Previously this would transform to `${PANTS_INTERNAL_ABSOLUTE_PREFIX}__something here` but 
the `startswith` would be false with the quoted version and so it'll come out as just `'__something here'`, incorrectly.

So... maybe move it within `prefixed` and be careful about where one uses the quoted vs. unquoted values:

```python
def prefixed_and_quote(arg: str) -> str:
    quoted = shlex.quote(arg)
    if arg.startswith("__"):
        return f"${{PANTS_INTERNAL_ABSOLUTE_PREFIX}}{quoted}"
    else:
        return quoted
```

NGustafson replied:
I've updated with that approach, thanks for all the feedback!

huonw reviewed (approved):
Nice, I like the test. I think we're close with the quoting, just a move of where it happens and we should be good to go!

Thanks for iterating!

kaos reviewed (approved):
thanks for your patience iterating on this :)

huonw reviewed (approved):
Nice one! Thanks the somewhat epic iteration!
